### ISSUE TITLE
Very slow performance when using Remote - WSL 

### ISSUE THREAD

**Issue by ahmedkrmn** [microsoft/vscode-cpptools#5558]
**Type: LanguageService**

**Describe the bug**
- OS and Version: Windows_NT 10.0.18363
- VS Code Version: 1.45.1
- C/C++ Extension Version: 0.28.1
    I tried downgrading to previous versions but the issue still persists.
- Does this issue involve using SSH remote to run the extension on a remote machine?: Yes

**Description:**
Code intellisense as well as auto formatting are very slow for very simple C++ files when using the Remote - WSL extension for C++ development on WSL.

![Screenshot (206)](https://user-images.githubusercontent.com/20889958/82751051-df0bfb80-9db4-11ea-8b37-eda1f9ca3ba5.png)

**`uname -r` output:**
4.4.0-18362-Microsoft

This issue isn't there when developing on Windows.

**Steps to reproduce**

Create a hello world C++ file and allow formatting on save from VScode.

**Comment by Colengms** [comment]
Hi @ahmedkrmn .  When remoting, the extension runs on the target machine.  Are you remoting into a slow system, such as an embedded device?

Regarding the performance of formatting; The C/C++ Extension currently uses clang-format for formatting.  A copy of clang-format is bundled with the extension, but a version installed on the (remote) device will be used instead if newer.  You might try running clang-format directly, on the device, to see if that time is being spent by it.  If so, perhaps there is a version of clang-format available that is optimized for your system.  If clang-format actually completes very quickly when run directly, let us know.  That might point to another issue.

**Comment by ahmedkrmn** [comment]
Hi @Colengms,
The target machine is an Ubuntu WSL version 4.4.0-18362-Microsoft.
I tried directly using clang-format as suggested and it formatted without any delays.

I did some further inspections and it appears that the issue is only preset when the workspace is the home directory of the WSL. When I created a new directory and used the VS remote extension to use it as the workspace, the issue disappears and formatting as well as linting is working well.

**Comment by sean-mcmanus** [comment]
I'm not reproing the issue. Is this with the "/home/" directory? It sounds like something is blocking our main input thread.

**Comment by MitchTalmadge** [comment]
I am having the same issues although my workspace is not the home directory. Autoformatting is very slow. How can I help provide info for debugging?

- OS and Version: 
  - Host: Windows 10 v2004
  - WSL: Debian 9 stretch
- VS Code Version: 1.46.1
- C/C++ Extension Version: 0.28.3

**Comment by sean-mcmanus** [comment]
@MitchTalmadge Are your files under /mnt? I think /mnt access can be slow with WSL2.

If you can wait for CPU usage to drop to 0 before doing a format, that would help isolate if the formatting is slow or just being blocked by previous processing (formatting is not expected to be slow, so probably other operations are blocking the formatting).

**Comment by MitchTalmadge** [comment]
> @MitchTalmadge Are your files under /mnt? I think /mnt access can be slow with WSL2.
> 
> If you can wait for CPU usage to drop to 0 before doing a format, that would help isolate if the formatting is slow or just being blocked by previous processing (formatting is not expected to be slow, so probably other operations are blocking the formatting).

They may be, I will have to double check and let you know. Thanks for the follow-up, really appreciate it

**Comment by ZijunHui** [comment]
I have a similar experience on WSL2 as well. My files are stored in /mnt/ and Intellisense is extremely slow, requiring more than 3 seconds for the results to pop up. After moving the files to ~, there is no longer a delay (or a very small one). I suspect this is due to WSL2, which has suprisingly slow access speed for /mnt/

**Comment by ghost** [comment]
I have a similar experience on WSL1. There's a CPU usage surge on cpptools:
![image](https://user-images.githubusercontent.com/32400590/95639693-4f5aa700-0a4e-11eb-879f-05642ae67126.png)

I also checked debug level log, it stayed at `Formatting Engine: clangFormat` for a while.

I tried to use clang-format directly from a terminal, it's fast.

The intellisnese updating and file analyzing speed is also slow.

**Comment by corneliusroemer** [comment]
Same issue here. All my files are under `/home`, I'm using WSL:Ubuntu-20.04. Intellisense is extremely slow - not usable. I'm using it for Python.

**Comment by pierce-martin** [comment]
Same issue for me. Files are under the home directory. Versions are:
- vscode: 1.50.1
- WSL: Ubuntu 18.04
- C/C++ extension: 1.0.1 (both local and remote)

Main issue I see is tab completion is slow and go to declaration/definition doesn't work. I see cpptools CPU usage spike as well.

**Comment by ghost** [comment]
> I have a similar experience on WSL1. There's a CPU usage surge on cpptools:
> ![image](https://user-images.githubusercontent.com/32400590/95639693-4f5aa700-0a4e-11eb-879f-05642ae67126.png)
> 
> I also checked debug level log, it stayed at `Formatting Engine: clangFormat` for a while.
> 
> I tried to use clang-format directly from a terminal, it's fast.
> 
> The intellisnese updating and file analyzing speed is also slow.

I upgraded my computer from a 2c4t VDI to a 16c16t VDI and switched from WSL1 to WSL2. The issue has been improved significantly. Though I'm not sure if it's due to CPU performance or WSL1 -> WSL2.

**Comment by stertingen** [comment]
I have the same issue when working on a ROS workspace. (ROS workspace consists of 3 directories: src, build and devel, with devel being some sort of install target location. There are many files in these directories, but not too many IMHO: 670 files in 192 directories)

VS Code 1.49.1
C/C++ extension: 1.1.2
ROS extension: 0.6.4
WSL: Ubuntu 20.04 (Linux 4.4.0-18362-Microsoft # 1049-Microsoft Thu Aug 14 12:01:00 PST 2020 x86_64 x86_64 x86_64 GNU/Linux)
Host: Windows 10 Enterprise version 1909 build 18363.1198

**Comment by Kraego** [comment]
I have the same issue with haskell and C/C++, files in home dir. Haskell language server on WSL.

I think this could be a wsl or code issue, since it's happening in c/c++, python, haskell, ...

VS Code 1.52.1
WSL2: Debian
Host: Windows 10 Home 1909

**Comment by klakiervx** [comment]
Same for me. cpptools sometimes hogs one full CPU core (~8% on 6-core with HT) for a couple of minutes (especially after opening another c++ file) or 10-30 seconds after modifying something in the file. IntelliSense popup often times out before it gets any data to show. Almost impossible to work like this.

VSCode 1.53.1
WSL1Ubuntu 18.04
Windows 10 Enterprise 1909

**Comment by rjkz808** [comment]
Same issue with WSL 1 and TypeScript Intellisense/type-checking when files are located in `/mnt/`.

**Comment by mmatrosov** [comment]
I have the same issue with real Ubuntu (not WSL). But I also connect via ssh, thus it is probably related.

VSCode:
```
Version: 1.54.3
Commit: 2b9aebd5354a3629c3aba0a5f5df49f43d6689f8
Date: 2021-03-15T10:55:24.277Z
Electron: 11.3.0
Chrome: 87.0.4280.141
Node.js: 12.18.3
V8: 8.7.220.31-electron.0
OS: Linux x64 5.8.0-44-generic snap
```
Host: Ubuntu 20.04.2 LTS
Remote: Ubuntu 20.04.2 LTS

**Comment by valerioformato** [comment]
> Same for me. cpptools sometimes hogs one full CPU core (~8% on 6-core with HT) for a couple of minutes (especially after opening another c++ file) or 10-30 seconds after modifying something in the file. IntelliSense popup often times out before it gets any data to show. Almost impossible to work like this.
> 
> VSCode 1.53.1
> WSL1Ubuntu 18.04
> Windows 10 Enterprise 1909

Same for me. I have the exact same symptoms 
Ubuntu 20.04 running on WSL (Win10 build 19041.487)
VSCode 1.54.3
C/C++ Extension Version: 1.2.2

**Comment by valerioformato** [comment]
I also tried to attach `strace` and then triggering one of these 100%cpu-burning moments on `cpptools`
What i see is that the process hangs for several seconds on a call to

**Comment by sean-mcmanus** [comment]
@valerioformato Are you able to locate the call stack that is calling the futex? Are those the reads in your log? If so, what calls precede the reads?

**Comment by valerioformato** [comment]
Unfortunately not, also the break signal doesn't get processed until the whole system unfreezes again :(

**Comment by valerioformato** [comment]
> There are instructions on how to get performance/profiler logging for Linux at https://github.com/microsoft/vscode-cpptools/wiki/Troubleshooting-Performance-Issues .

Doesn't look like perf can run in WSL 1 :\

**Comment by sean-mcmanus** [comment]
Have you tried the Windows profiling approach? With WSL 1 the processes may still be considered Windows processes (while with WSL2 they run in some sort of Linux VM).

**Comment by valerioformato** [comment]
> Have you tried the Windows profiling approach? With WSL 1 the processes may still be considered Windows processes (while with WSL2 they run in some sort of Linux VM).

Ouch, no I didn't think about that. I had been postponing the upgrade to WSL2 due to several issues with virtualization on my system. (that was also preventing me from using docker)
Lately the frustration has driven me to spend the afternoon fighting with that, and I managed to finally upgrade. I'll let you know if the issue persists.

**Comment by mvoloskov** [comment]
So, if I have Home Edition and thus no WSL 2.0, I'm doomed to slow performance on Windows?

**Comment by sean-mcmanus** [comment]
@mvoloskov If you search for "Windows Home edition wsl2" you should be able to find some steps to enable WSL 2.0 on Home Edition -- let us know if that doesn't work.

**Comment by mvoloskov** [comment]
Thank you all, I ended up just using Git for Windows and uninstalling WSL completely.

**Comment by sbland** [comment]
I have the same issue but only when I also get the **Workspace is too large to track changes** message. The workspace isn't a git repo but contains a python venv directory that is presumably too large to track. 
I fixed the problem by adding `**/venv/**` to the watcher ignore list: 
``` json
"files.watcherExclude": {
  "**/venv/**": true
}
```

**Comment by chriszhao001** [comment]
Same for me. But the main process which make whole PC very slow in windows task manager in "node" process.
When I use visual studio code open an directory in WSL "home" directory, I can see an process "node" in task manager.
![image](https://user-images.githubusercontent.com/17903065/131433118-1d492cdc-c8b3-489c-8b6a-c6d0f7aba2b7.png)
after I close visual studio or move to another directory which is not in WSL "home" directory, the "node" process disappear.
the node file is located in /home/xxx/.vscode-server/bin/3866c3553be8b268c8a7f8c0482c0c0177aa8bfa

Ubuntu 20.04 running on WSL (Win10 build 19042.1165)
VSCode 1.59.1
C/C++ Extension Version: 1.6.0

**Comment by Veenkar** [comment]
How about: crtl+shift+P -> reset intellisense database?

**Comment by mort3za** [comment]
I had this problem. The reason was moving the project folder from Windows file system to Ubuntu file system without fixing file permissions.
Cloned my project again and it fixed the issue.

**Comment by adams-family** [comment]
Is it possible to use VS Code locally (all extensions locally) and only edit files on the remote machine?

It's is an issue for me, as well, on all cheap VPS servers.

**Comment by michelleangela** [comment]
@adams-family 
>Is it possible to use VS Code locally (all extensions locally) and only edit files on the remote machine?

**Short answer:**
In order for the language server to effectively parse a project for IntelliSense, it needs to run on the same machine where the project and compiler are on.

**Example:**
The language server requires parsing of all files in a project in order to find and process symbols. If a project is on a remote machine, then the entire project as well as system header dependencies will need to be sent over to the local machine where the language server is running. Also, the extension queries the compiler in order to find system headers and get defines.

**Comment by adams-family** [comment]
@michelleangela Got it, makes sense! In this case, can I work with VS Code locally (on my laptop, including IntelliSense) and sync my files with the remote server on each Save?

**Comment by adams-family** [comment]
@michelleangela Understood. I'm administering a lot of small projects where I occasionally need to connect, change a configuration file (or a source code), debug the problem and leave. The issue is when VS code server-extensions eat out the resources on the remote VPS, it becomes unresponsive. I used NetBean's remote sync for this but that project got abandoned. Thanks for the clarification anyway!

**Comment by grapereader** [comment]
Similar issue... formatting C++ files never completes on a remote host. The host has ample RAM and CPU, but everything (including vscode-server itself) is on a very slow network share.

Here's some speculation:  On this system, disk reads are fast enough, but disk writes are slow. Especially with lots of small files. VSCode must not be optimized for this (is it shuffling around small intermediate files?). This would also explain poor performance with WSL1 which struggles with disk speed.

**Comment by ginkgo97** [comment]
Similar issue. I installed wsl1 on my win10 os.
- vscode version: 1.69.0
- wsl1 version: ubuntu 22.04
- c/c++ extension version: 1.10.8 (on wsl only)
When I enable the Remote-WSL plugin and connect to wsl, it is slowly responsed if I use bash on wsl within vscode.
But when I disconnect from wsl, and only use bash on wsl within vscode, everything is ok.

**Comment by ctjlewis** [comment]
A team member of mine switched to Mac to workaround. This issue should be prioritized based on the number of user reports.

**Comment by luciandex** [comment]
Yes, the problem is still alive. 
Indexing with intellisense php code, it still run for 24 hours to index a Magento 2 installation in WSL2.

**Comment by nathanpovo** [comment]
Getting the same issue with `prettier` when developing a react typescript project.

### System info

- Remote: WSL: Ubuntu
- OS: Linux x64 5.15.90.1-microsoft-standard-WSL2
- VS Code Version: 1.76.0

### Relevant extension info

- prettier-vscode version 9.10.4

**Comment by Colengms** [comment]
There's been a lot of traffic on this issue, which suggests many people are experiencing perf issues when using the C/C++ Extension remotely via WSL.

The C/C++ Extension runs natively on a 'remote' system.  It does not operate differently specifically for WSL remoting.  It's native processes are the same as used on other native Linux platforms.  While it's possible that there could be something the C/C++ Extension is doing that could be less efficient on Linux than on Windows, it's likely that performance issues in WSL are specific to VS Code's WSL remoting, or WSL itself.  For example, if a project is on the Windows file system and accessed through a `/mnt` mount point, file operations are going through an additional layer which may account for some performance impact.  It's less efficient than the same workspace opened on Windows.  In the past, we have seen WSL bugs specifically with `/mnt` mount points.

For many WSL performance issues, the resolution would be unlikely to come from the C/C++ Extension.  Consider investigating whether similar performance issues can be reproduced with VS Code itself or other extensions, and open issues in [the VS Code repo](https://github.com/microsoft/vscode/issues), if so.

**Comment by DarrenLevine** [comment]
I found a workaround for this issue:

1. Install the [jkillian.custom-local-formatters](https://marketplace.visualstudio.com/items?itemName=jkillian.custom-local-formatters) extension
2.  Add the following lines to your /.vscode/settings.json file:
```json
"customLocalFormatters.formatters": [
    {
        "command": "clang-format",
        "languages": [
             "cpp"
        ]
    }
  ]
```
3. Open any C++ file and run "> Format Document" in VSCode's command palette.
4. Vscode will ask you which formatter to use, select the custom-local-formatters option as the new default.
5. From then on, when you save a C++ document, it will ignore the slower C++ extension's default formatter, and use clang-format directly, which is much faster.

Note that you can also use a ".clang-format" file with this method, and set the BasedOnStyle option in that file to LLVM to get a format similar to what the vscode C++ extension's formatter uses as a default configuration.

**Comment by sean-mcmanus** [comment]
As I believe I mentioned earlier, this is caused by our formatting operating getting stuck behind an IntelliSense update. It's possible this could be fixed or improved by https://github.com/microsoft/vscode-cpptools/issues/7759 .
